File: checkout/src/js/http/ContactApi.ts
import {HttpService} from "../common/services/HttpService";
import {ConcessionsAndMemberships} from "../model/checkout/ConcessionsAndMemberships";
import {CreateParentChildrenRequest} from "../model/checkout/CreateParentChildrenRequest";
import {Concession} from "../model/checkout/concession/Concession";
import {ConcessionType} from "../model/checkout/concession/ConcessionType";
import {ContactFields} from "../model/field/ContactFields";
import {ContactFieldsRequest} from "../model/field/ContactFieldsRequest";
import {Contact} from "../model/web/Contact";
import {ContactId} from "../model/web/ContactId";
import {CreateContactParams} from "../model/web/CreateContactParams";

export class ContactApi {
  constructor(private http: HttpService) {
  }

  createOrGetContact(createContactParams: CreateContactParams): Promise<ContactId> {
    return this.http.PUT(`/v1/contact`, createContactParams);
  }
  createParentChildrenRelation(request: CreateParentChildrenRequest): Promise<any> {
    return this.http.PUT(`/v1/createParentChildrenRelation`, request);
  }
  getConcessionTypes(): Promise<ConcessionType[]> {
    return this.http.GET(`/v1/getConcessionTypes`);
  }
  getContact(studentUniqueIdentifier: string): Promise<Contact> {
    return this.http.GET(`/v1/contact/${studentUniqueIdentifier}`);
  }
  getContactConcessionsAndMemberships(contactIds: string[]): Promise<ConcessionsAndMemberships> {
    return this.http.POST(`/v1/getContactConcessionsAndMemberships`, contactIds);
  }
  getContactFields(contactFieldsRequest: ContactFieldsRequest): Promise<ContactFields> {
    return this.http.POST(`/v1/contactFields`, contactFieldsRequest);
  }
  submitConcession(concession: Concession): Promise<any> {
    return this.http.PUT(`/v1/submitConcession`, concession);
  }
  submitContactDetails(contactFields: ContactFields): Promise<ContactId> {
    return this.http.POST(`/v1/submitContactDetails`, contactFields);
  }
}

File: checkout-api/src/main/groovy/ish/oncourse/willow/service/ContactApi.java
package ish.oncourse.willow.service;

import java.util.List;
import ish.oncourse.willow.model.checkout.ConcessionsAndMemberships;
import ish.oncourse.willow.model.checkout.CreateParentChildrenRequest;
import ish.oncourse.willow.model.checkout.concession.Concession;
import ish.oncourse.willow.model.checkout.concession.ConcessionType;
import ish.oncourse.willow.model.field.ContactFields;
import ish.oncourse.willow.model.field.ContactFieldsRequest;
import ish.oncourse.willow.model.web.Contact;
import ish.oncourse.willow.model.web.ContactId;
import ish.oncourse.willow.model.web.CreateContactParams;

import javax.ws.rs.*;

@Path("/")
public interface ContactApi  {

    @PUT
    @Path("/v1/contact")
    @Produces({ "application/json" })
    @XValidate
    @CollegeInfo
    @ContactValidation
    ContactId createOrGetContact(CreateContactParams createContactParams);

    @PUT
    @Path("/v1/createParentChildrenRelation")
    @Consumes({ "application/json" })
    @Produces({ "application/json" })
    @XValidate
    @CollegeInfo
    void createParentChildrenRelation(CreateParentChildrenRequest request);

    @GET
    @Path("/v1/getConcessionTypes")
    @Produces({ "application/json" })
    @CollegeInfo
    List<ConcessionType> getConcessionTypes();

    @GET
    @Path("/v1/contact/{studentUniqueIdentifier}")
    @Produces({ "application/json" })
    @CollegeInfo
    Contact getContact(@PathParam("studentUniqueIdentifier") String studentUniqueIdentifier);

    @POST
    @Path("/v1/getContactConcessionsAndMemberships")
    @Produces({ "application/json" })
    @CollegeInfo
    ConcessionsAndMemberships getContactConcessionsAndMemberships(List<String> contactIds);

    @POST
    @Path("/v1/contactFields")
    @Consumes({ "application/json" })
    @Produces({ "application/json" })
    @CollegeInfo
    ContactFields getContactFields(ContactFieldsRequest contactFieldsRequest);

    @PUT
    @Path("/v1/submitConcession")
    @Produces({ "application/json" })
    @XValidate
    @CollegeInfo
    void submitConcession(Concession concession);

    @POST
    @Path("/v1/submitContactDetails")
    @Consumes({ "application/json" })
    @Produces({ "application/json" })
    @XValidate
    @CollegeInfo
    ContactId submitContactDetails(ContactFields contactFields);
}

